src/multiselect.js

```diff
@@ -0,0 +1,206 @@
+/**
+ * @license
+ * Copyright 2022 MIT
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+/**
+ * @fileoverview Plugin main class.
+ */
+
+import * as Blockly from 'blockly/core';
+
+import * as ContextMenu from './multiselect_contextmenu';
+import * as Shortcut from './multiselect_shortcut';
+import {blockSelection, inMultipleSelectionMode,
+  hasSelectedParent, setBaseBlockDragger} from './global';
+import {MultiselectControls} from './multiselect_controls';
+
+/**
+ * Class for using multiple select blocks on workspace.
+ */
+export class Multiselect {
+  /**
+   * Initalize the class data structure.
+   * @param {!Blockly.WorkspaceSvg} workspace The workspace to sit in.
+   */
+  constructor(workspace) {
+    this.workspace_ = workspace;
+    this.origHandleWsStart_ = Blockly.Gesture.prototype.handleWsStart;
+  }
+
+  /**
+   * Bind the events and replace registration.
+   * @param {{multiselectIcon: Object, useDoubleClick: boolean}} options
+   * to set.
+   */
+  init(options) {
+    this.onKeyDownWrapper_ = Blockly.browserEvents.conditionalBind(
+        this.workspace_.getInjectionDiv(), 'keydown', this, this.onKeyDown_);
+    this.onKeyUpWrapper_ = Blockly.browserEvents.conditionalBind(
+        this.workspace_.getInjectionDiv(), 'keyup', this, this.onKeyUp_);
+    this.eventListenerWrapper_ = this.eventListener_.bind(this);
+    this.workspace_.addChangeListener(this.eventListenerWrapper_);
+
+    ContextMenu.unregisterContextMenu();
+    ContextMenu.registerOurContextMenu();
+    Shortcut.unregisterShortcut();
+    Shortcut.registerOurShortcut();
+
+    this.controls_ = new MultiselectControls(
+        this.workspace_, options.multiselectIcon, this);
+    if (!options.multiselectIcon || !options.multiselectIcon.hideIcon) {
+      const svgControls = this.controls_.createDom();
+      this.workspace_.getParentSvg().appendChild(svgControls);
+    }
+    this.controls_.init();
+
+    if (options.useDoubleClick) {
+      this.useDoubleClick_(true);
+    }
+
+    if (options.baseBlockDragger) {
+      setBaseBlockDragger(options.baseBlockDragger);
+    }
+
+    if (!options.bumpNeighbours) {
+      this.origBumpNeighbours = Blockly.BlockSvg.prototype.bumpNeighbours;
+      Blockly.BlockSvg.prototype.bumpNeighbours = function() {};
+    }
+  }
+
+  /**
+   * Unbind the events and replace with original registration.
+   */
+  dispose() {
+    if (this.onKeyDownWrapper_) {
+      Blockly.browserEvents.unbind(this.onKeyDownWrapper_);
+      this.onKeyDownWrapper_ = null;
+    }
+    if (this.onKeyUpWrapper_) {
+      Blockly.browserEvents.unbind(this.onKeyUpWrapper_);
+      this.onKeyUpWrapper_ = null;
+    }
+    if (this.eventListenerWrapper_) {
+      this.workspace_.removeChangeListener(this.eventListenerWrapper_);
+      this.eventListenerWrapper_ = null;
+    }
+
+    ContextMenu.unregisterContextMenu();
+    ContextMenu.registerOrigContextMenu();
+    Blockly.ContextMenuRegistry.registry.unregister('workspaceSelectAll');
```

Comment:
It seems like it would be better to group all of the unregister calls together followed by all of the register calls. Having this here could undo a registration that happened in the call on the previous line.